### PR TITLE
servlets: link access counter

### PR DESCRIPTION
**Pull request by slavidichkov**
TBD

### CODE REVIEW

File: src/main/java/com/clouway/task1/http/LinkAccessCounter.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.PrintWriter;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class LinkAccessCounter extends HttpServlet {
+    private int firstLinkCounter = 0;
```
Comment (mgenov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: We already discussed this on presentation. It's not thread safe cause Servlets are singleton objects and these variables are going to be re-used between requests.

File: src/main/java/com/clouway/task1/http/LinkAccessCounter.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    private int secondLinkCounter = 0;
+    private int thirdLinkCounter = 0;
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        String linkId = req.getParameter("linkId");
```
Comment (mgenov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: Session or Application scope could be used for the keeping of counter values.

File: src/test/java/com/clouway/task1/http/LinkAccessCounterTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            oneOf(response).getWriter();
+            will(returnValue(new PrintWriter(out)));
+        }});
+        linkAccessCounter.doGet(request, response);
+        String expected=out.toString();
+        assertThat(expected.contains("SecondLink 1"),is(true));
+    }
+
+    @Test
+    public void thirdLinkCounter() throws ServletException, IOException {
```
Comment (mgenov) [inline src/test/java/com/clouway/task1/http/LinkAccessCounterTest.java]: If we follow your test naming convention, then this test should be named: oneInvocationOfThirdLink probably.

Comment (slavidichkov) [inline src/test/java/com/clouway/task1/http/LinkAccessCounterTest.java]: fixed

File: src/main/java/com/clouway/task1/http/LinkAccessCounter.java (hunk 1 above)
Comment (slavidichkov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: Now my counter  is common for all users. I did not understand that counter should be for each user.

Comment (slavidichkov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: I will fixed

Comment (slavidichkov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: fixed

File: src/main/java/com/clouway/task1/http/LinkAccessCounter.java (hunk 2 above)
Comment (slavidichkov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: done

**Comment by mgenov** [comment]
Could you add another counter, to see what will happen with the code ?

File: src/main/java/com/clouway/task1/http/LinkAccessCounter.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class LinkAccessCounter extends HttpServlet {
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        final String linkId = req.getParameter("linkId");
+        HttpSession session = req.getSession();
```
Comment (mgenov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: This code looks messy and requires some cleanup.

Comment (slavidichkov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: if I add another one counter i have to add one more parameters in my print method and one more if statement. I can put in map all counters for key with values 0 and my code will be more clear

Comment (mgenov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: So you have to change the existing code? 

If so, then the current code is breaking the open closed principle which states: "software entities (classes, modules, functions, etc.) should be open for extension, but closed for modification"

Comment (slavidichkov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: I pass to the constructor Map with page name for key and value 0.

Comment (mgenov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: What happens if 2 users click on the same time on link 1 ?

Comment (slavidichkov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: by that way i wrote in the same map for all users

Comment (slavidichkov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: fixed

Comment (mgenov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: Does not seem so.

Comment (mgenov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: As your objects in map are String and Integer, it's okay.

Comment (slavidichkov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: now i saved map in session atribute for each user

Comment (mgenov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: Yeah. I'm fine with what I saw.

File: src/main/java/com/clouway/task1/http/LinkAccessCounter.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            Integer counter = linkCounters.get(linkId);
+            linkCounters.put(linkId, counter + 1);
+        }
+        PrintWriter writer = resp.getWriter();
+        writer.println(getHtmlContent(linkCounters));
+        writer.flush();
+    }
+
+    private String getHtmlContent(Map<String, Integer> linkCounters) {
+        StringBuilder sb=new StringBuilder();
```
Comment (tsony-tsonev) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: sb is not a good name

File: src/main/java/com/clouway/task1/server/Jetty.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    private final Server server;
+
+    public Jetty(int port) {
+        this.server = new Server(port);
+    }
+
+    public void start() {
+        ServletContextHandler context = new ServletContextHandler(ServletContextHandler.SESSIONS);
+        context.setContextPath("/");
+        context.addEventListener(new ServletContextListener() {
```
Comment (tsony-tsonev) [inline src/main/java/com/clouway/task1/server/Jetty.java]: You can extract this ServletContextListener in separate class with proper name to make the code more readable. This way you'll have a class with the servlet bindings and here you'll just add the listeners.

**Comment by tsony-tsonev** [comment]
LGTM... you can merge this after fixing the last comments

File: src/main/java/com/clouway/task1/server/Jetty.java (hunk 6 above)
Comment (slavidichkov) [inline src/main/java/com/clouway/task1/server/Jetty.java]: fixed

File: src/main/java/com/clouway/task1/http/LinkAccessCounter.java (hunk 5 above)
Comment (slavidichkov) [inline src/main/java/com/clouway/task1/http/LinkAccessCounter.java]: fixed

**Pull request merged**